eZ/Publish/Core/Repository/RoleService.php

```diff
@@ -881,25 +877,58 @@ public function getRoleAssignments(APIRole $role): iterable
             throw new UnauthorizedException('role', 'read');
         }
 
+        $persistenceRoleAssignments = $this->userHandler->loadRoleAssignmentsByRoleId($role->id);
+
+        return $this->buildRoleAssignmentsFromPersistence($role, $persistenceRoleAssignments);
+    }
+
+    /**
+     * @return \eZ\Publish\API\Repository\Values\User\RoleAssignment[]
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to read a role
+     */
+    public function loadRoleAssignments(APIRole $role, int $offset = 0, int $limit = -1): iterable
+    {
+        if (!$this->permissionResolver->canUser('role', 'read', $role)) {
+            throw new UnauthorizedException('role', 'read');
+        }
+
+        $persistenceRoleAssignments = $this->userHandler->loadRoleAssignmentsByRoleIdWithOffsetAndLimit(
+            $role->id,
+            $offset,
+            $limit
+        );
+
+        return $this->buildRoleAssignmentsFromPersistence($role, $persistenceRoleAssignments);
+    }
+
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     */
+    private function buildRoleAssignmentsFromPersistence(
+        APIRole $role,
+        array $persistenceRoleAssignments
+    ): array {
```

Comment:
Given the consumers of that method return `iterable`, how about making this `iterable` as well by yielding the results rather than accumulating them in the `$roleAssignments` array?

Reply:
@alongosz the problem is that even if consumers return `iterable`, some places in other packages expect those consumer methods to return a simple array.

Reply:
@alongosz sure, they return `iterable`, but in fact, both methods return simple arrays, even before my change. For example, an array is required here https://github.com/ezsystems/ezplatform-admin-ui/blob/2.3/src/bundle/Controller/RoleAssignmentController.php#L241. So what I meant by my previous comment is that we would have to convert those `Generator` yield values to a simple array in `getRoleassignments` and `loadRoleAssignments` with `iterator_to_array` or something similar.

Reply:
@alongosz basically while our contract specifies we will return iterables, in a few places (in our own code) we operate as if it was always array.

Unless we can identify and fix all those places (including any potential usage of these in third-party code) we should refrain from changing it, especially in 3.x codebase.

Reply:
Seems rather like a consumer bug, but fair enough, if the impact is high, let's keep it as-is :+1: